cocotb/regression.py

```diff
@@ -229,63 +230,112 @@ def next_test(self):
             return None
         return self._queue.pop(0)
 
-    def _add_failure(self, result):
-        self.xunit.add_failure(stdout=repr(str(result)),
-                               stderr="\n".join(self._running_test.error_messages),
-                               message="Test failed with random_seed={}".format(self._seed))
-        self.failures += 1
-
     def handle_result(self, test):
         """Handle a test completing.
 
-        Dump result to XML and schedule the next test (if any).
+        Dump result to XML and schedule the next test (if any). Entered by the scheduler.
 
         Args:
             test: The test that completed
         """
         assert test is self._running_test
 
-        real_time   = time.time() - test.start_time
+        real_time = time.time() - test.start_time
         sim_time_ns = get_sim_time('ns') - test.start_sim_time
-        ratio_time  = self._safe_divide(sim_time_ns, real_time)
-        
+        ratio_time = self._safe_divide(sim_time_ns, real_time)
+
         self.xunit.add_testcase(name=test.funcname,
                                 classname=test.module,
                                 time=repr(real_time),
                                 sim_time_ns=repr(sim_time_ns),
                                 ratio_time=repr(ratio_time))
 
+        # check what exception the test threw
+        try:
+            test._outcome.get()
+        except Exception as e:
+            result_pass, sim_failed = self.score_test(test, *self.exception_info(e, ['handle_result', 'get']))
+        else:
+            result_pass, sim_failed = self.score_test(test, TestSuccess(), None)
+
+        # Save results
+        self._store_test_result(test.__module__, test.__name__, result_pass, sim_time_ns, real_time, ratio_time)
+        if not result_pass:
+            self.xunit.add_failure()
+            self.failures += 1
+
+        # Fail if required
+        if sim_failed:
+            self.tear_down()
+
+        self.execute()
+
+    @staticmethod
+    def exception_info(e, frame_names):
```

Comment:
Instead of taking e and frame_names, could we pass in an outcome here? It would make the interface a little simpler, and remove the restriction that this method can only be called in an except block.

Reply:
True. I removed the function and changed `score_test` to take an outcome. I noticed a potential improvement on `Outcome` would be to add a `__bool__` method for checking whether the outcome is an `Error` or a `Value`. Seeing as `Outcome` is basically an option monad, its natural for people to use it that way.

Reply:
I removed this function and changed `score_test` to take an outcome.